monix-tail/shared/src/main/scala/monix/tail/internal/IterantDropWhileWithIndex.scala

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2014-2018 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.tail.internal
+
+import cats.effect.Sync
+import cats.syntax.all._
+import monix.execution.misc.NonFatal
+import monix.tail.Iterant
+import monix.tail.Iterant.{Halt, Last, Next, NextBatch, NextCursor, Suspend}
+import monix.tail.batches.BatchCursor
+import scala.annotation.tailrec
+
+private[tail] object IterantDropWhileWithIndex {
+  /**
+    * Implementation for `Iterant#dropWhileWithIndex`
+    */
+  def apply[F[_], A](source: Iterant[F, A], p: (A, Int) => Boolean)
+                    (implicit F: Sync[F]): Iterant[F, A] = {
+
+    // Reusable logic for NextCursor / NextBatch branches
+    @tailrec
+    def evalCursor(ref: F[Iterant[F, A]], cursor: BatchCursor[A], rest: F[Iterant[F, A]],
+                   stop: F[Unit], dropped: Int, index: Int, continueDropping: Boolean): Iterant[F, A] = {
+      if (!cursor.hasNext())
+        Suspend(rest.map(loop(index, continueDropping)), stop)
+      else if (dropped >= cursor.recommendedBatchSize)
+        Suspend(ref.map(loop(index, continueDropping)), stop)
+      else {
+        val elem = cursor.next()
+        if (continueDropping && p(elem, index))
+          evalCursor(ref, cursor, rest, stop, dropped + 1, index + 1, continueDropping)
+        else if (cursor.hasNext())
+          Next(elem, ref.map(loop(index, continueDropping = false)), stop)
+        else
+          Next(elem, rest.map(loop(index, continueDropping = false)), stop)
+      }
+    }
+
+    def loop(index: Int, continueDropping: Boolean)(source: Iterant[F, A]): Iterant[F, A] = {
+      try source match {
+        case ref@Next(item, rest, stop) =>
+          if (continueDropping && p(item, index)) Suspend(rest.map(loop(index + 1, continueDropping = true)), stop)
+          else ref
```

Comment:
Another personal pet peeve — that line is too long, so something like this maybe?

```scala
if (continueDropping && p(item, index)) 
  Suspend(rest.map(loop(index + 1, continueDropping = true)), stop)
else 
  ref 
```